Src/ILGPU.Algorithms/MatrixOperations/MaskedSparseMatrixExtensions.cs

```diff
@@ -0,0 +1,448 @@
+        /// <param name="outView">
+        /// A dense output view containing the results of the multiplication.
+        /// </param>
+        /// <param name="processor">An instance of an actual mul-add operation.</param>
+        internal static void MaskedSparseTransposedMatrixMultiplierKernel<
+            T,
+            TPredicate,
+            TStride,
+            TProcessor>(
+            TPredicate maskPredicate,
+            ArrayView2D<T, TStride> aView,
+            SparseMatrixView<T, TStride> bView,
+            ArrayView2D<T, TStride> outView,
+            TProcessor processor)
+            where T : unmanaged
+            where TStride : struct, IStride2D
+            where TPredicate : struct, InlineList.IPredicate<Index2D>
+            where TProcessor : struct, IMaskedSparseMatrixProcessor<T>
+        {
+            var maskView = SharedMemory.GetDynamic<int>();
+            Trace.Assert(
+                Group.DimX <= maskView.IntLength,
+                "Invalid shared memory config");
+
+            // Get all predicate results and perform thread compaction
+            int index = Grid.GlobalLinearIndex;
+            var maskIndex2D = outView.Stride.ReconstructFromElementIndex(index);
+            bool apply =
+                maskIndex2D.X < outView.Extent.X &
+                maskIndex2D.Y < outView.Extent.Y;
+
+            // Check the predicate (if in range) and compute thread-compaction offsets
+            if (apply)
+                apply = maskPredicate.Apply(maskIndex2D);
+            int groupOffset = GroupExtensions.ExclusiveScan<int, AddInt32>(
+                Utilities.Select(apply, 1, 0));
+
+            // Store current result and determine the total number of matrix entries
+            if (apply)
+                maskView[groupOffset] = index;
+            int numEntries = Group.BarrierPopCount(apply);
+
+            // Load the current entry (if any)
+            if (Group.IdxX >= numEntries)
+                return;
+
+            // Get the transposed index from our processing index
+            int processingIndex = maskView[Group.IdxX];
+            var index2D = outView.Stride.ReconstructFromElementIndex(
+                processingIndex);
+
+            // Load next number of neighbors
+            T dotProduct = default;
+            int numNeighbors = bView.NumNeighbors[index2D.Y];
+
+            // Note that profiling yielded that explicit L1 shared-memory caching did not
+            // help to improve performance in this case. This is also due to the fact
+            // that most reused results automatically end up being fetching into L2 cache
+            // and loaded upon request (based on the structure of the algorithm) into the
+            // L1 parts of the processing unit
+
+            for (var neighborIndex = 0; neighborIndex < numNeighbors; ++neighborIndex)
+            {
+                // Load index and sparse edge weight
```

Comment:
// Note that the format for bView means we will read the transposed entry from bView